RavenDB-12151

arekpalinski opened the pull request.

arekpalinski committed:
RavenDB-12151 Fixing counting the number of pages modified in a transaction

arekpalinski committed:
RavenDB-12151 Added a configuration option that will limit the amount of scratch space that we allow to use for an indexing batch

arekpalinski committed:
RavenDB-12151 Let's decrease amount of allowed scratch space usage for map-reduce indexes to take into account additional work in the reduce phase

arekpalinski committed:
RavenDB-12151 Added option to limit the amount of scratch space that is used in indexing globally

arekpalinski committed:
RavenDB-12151 Rename IScratchSpaceUsageMonitor to IScratchSpaceMonitor

arekpalinski committed:
RavenDB-12151 Include decompression buffers in scratch space usage

grisha-kotler commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -2940,6 +2985,24 @@ private int MinBatchSize
 
+            if (ScratchSpaceLimitInBytes != null && txAllocations > ScratchSpaceLimitInBytes)
+            {
+                stats.RecordMapCompletedReason($"Reached scratch space limit ({ScratchSpaceLimitInBytes / 1024:#,#0} kb). Allocated {txAllocations / 1024:#,#0} kb of scratch space in current transaction");
```
Can you use size here?

arekpalinski replied:
Sure. I did it this way for consistency with existing messages. Will fix it since I saw you changed it it another PR

grisha-kotler commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -2940,6 +2985,24 @@ private int MinBatchSize
+            {
+                _globalIndexingScratchSpaceLimitExceeded = true;
+
+                stats.RecordMapCompletedReason(
```
Same here

grisha-kotler commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -2940,6 +2985,24 @@ private int MinBatchSize
+                _globalIndexingScratchSpaceLimitExceeded = true;
+
+                stats.RecordMapCompletedReason(
+                    $"Reached global scratch space limit for indexing ({globalIndexingScratchSpaceUsage.LimitInBytes / 1024:#,#0} kb). Current scratch space is {globalIndexingScratchSpaceUsage.ScratchSpaceInBytes / 1024:#,#0} kb");
```
Same here

arekpalinski committed:
RavenDB-12151 Fixing PR comments

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -123,6 +124,27 @@ public virtual bool RunInMemory
+        [DefaultValue(null)]
+        [SizeUnit(SizeUnit.Megabytes)]
+        [IndexUpdateType(IndexUpdateType.Refresh)]
+        [ConfigurationEntry("Indexing.ScratchSpaceLimitInMb", ConfigurationEntryScope.ServerWideOrPerDatabase)]
```
can this be defined per index?

arekpalinski replied:
Yes. It can be defined per index as setting any other index specific configuration using `IndexConfiguration` class in client API via `IndexDefinition.Configuration`. It's also exposed in the studio.

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -123,6 +124,27 @@ public virtual bool RunInMemory
+        public Size? GlobalScratchSpaceLimit { get; protected set; }
+
+        [Description("Max time to wait when forcing the environment flush and sync after exceeding global scratch buffer limit")]
+        [DefaultValue(5)]
```
This should probably be a higher value by default.
We are in a bad place if this is hit, so we want to give the entire system more time to process things one at a time

arekpalinski replied:
15 seconds would be enough as the default?

ayende replied:
Lets go with 30 or so, just to be safer

ayende commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -2940,6 +2985,24 @@ private int MinBatchSize
+                stats.RecordMapCompletedReason($"Reached scratch space limit ({ScratchSpaceLimit.Value}). Allocated {new Size(txAllocations, SizeUnit.Bytes)} of scratch space in current transaction");
+                return false;
+            }
+
```
I think we are also missing another limit here. The total amount of scratch page used in this index.
The `txAllocations` is just for the current batch, if we didn't flush yet from the scratch files, we may be gathering a lot of them, not on a single batch basis, but across many of them.
We need a limit on the amount of scratch files we have

arekpalinski replied:
Good point. I'll rename this config to something that will state it's about max transaction size per _batch_. 

Then current `Indexing.ScratchSpaceLimitInMb` will track the scratch space using the introduced `IScratchSpaceMonitor` but on index environment level.

 If we hit this limit then I assume we should behave the same like hitting `Indexing.GlobalScratchSpaceLimitInMb` - flush and force sync, right?

ayende replied:
Yes

arekpalinski replied:
Fixed in https://github.com/ravendb/ravendb/pull/7881/commits/c017182f974b2fe0f6eff87f7549b7af16caf381

arekpalinski committed:
RavenDB-12151
- Added Indexing.TransactionSizeLimitInMb option that will stop the indexing batch based on current tx allocations
- Option Indexing.ScratchSpaceLimitInMb will limit the total amount of scratch space per index (not globally)
- Changed default from 5 to 30 seconds for Indexing.MaxTimeToWaitAfterFlushAndSyncWhenExceedingScratchSpaceLimit option

ppekrol commented:
LGTM. @ayende ?

ayende commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -1120,6 +1125,21 @@ protected void ExecuteIndexing()
+                                    _logger.Info($"Indexing exceeded global limit for scratch space usage. Going to flush environment of '{Name}' index and forcing sync of data file");
+
+                                GlobalFlushingBehavior.GlobalFlusher.Value.MaybeFlushEnvironment(storageEnvironment);
+                                GlobalFlushingBehavior.GlobalFlusher.Value.ForceSyncEnvironment(storageEnvironment);
```
Don't we need to wait until the flushing is done before doing the sync?

arekpalinski replied:
`MaybeFlushEnvironment` and `ForceSyncEnvironment` just enqueue environments for relevant operations The global flusher processes the flush operations first, next it do the syncs. So we're guaranteed that sync will be processed after the flush.

ayende replied:
Okay, then LGTM
